src/server/services/procedures/salio/salio.js

```diff
@@ -0,0 +1,749 @@
+ * It allows your android device to be used as a sensor which is accessible from inside Netsblox.
+ * 
+ * @alpha
+ * @service
+ */
+
+/*
+ * Based on the RoboScape procedure.
+ *
+ * Sensor to server messages:
+ *  mac_addr[6] time[4] 'I': identification, sent every second
+ *  mac_addr[6] time[4] 'A' x[4] y[4] z[4]: accelerometer data
+ *
+ * Server to sensor messages:
+ *  'A': get accelerometer snapshot
+ *
+ * Environment variables:
+ *  SALIO_PORT: set it to the UDP port (1975) to enable this module
+ *  SALIO_MODE: sets the NetsBlox interface type, can be "security",
+ *      "native" or "both" (default)
+ */
+
+'use strict';
+
+const {PromiseSocket} = require('promise-socket');
+
+const logger = require('../utils/logger')('SalIO');
+const Sensor = require('./sensor');
+const acl = require('../roboscape/accessControl');
+const common = require('./common');
+var dgram = require('dgram'),
+    server = dgram.createSocket('udp4'),
+    SALIO_MODE = process.env.SALIO_MODE || 'both';
+
+const SALIO_TCP_PORT = 8889; // tcp port used by the SalIO app
+
+/*
+ * SalIO - This constructor is called on the first
+ * request to an RPC from a given room.
+ * @constructor
+ * @return {undefined}
+ */
+var SalIO = function () {
+    this._state = {
+        registered: {}
+    };
+};
+
+SalIO.serviceName = 'SalIO';
+// keeps a dictionary of sensor objects keyed by mac_addr
+SalIO.prototype._sensors = {};
+
+SalIO.prototype._ensureLoggedIn = function() {
+    if (this.caller.username !== undefined)
+        throw new Error('Login required.');
+};
+
+SalIO.prototype._ensureAuthorized = async function(sensorId) {
+    await acl.ensureAuthorized(this.caller.username, sensorId);
+};
+
+// fetch the sensor and updates its address. creates one if necessary
+SalIO.prototype._getOrCreateSensor = function (mac_addr, ip4_addr, ip4_port) {
+    var sensor = this._sensors[mac_addr];
+    if (!sensor) {
+        logger.log('discovering ' + mac_addr + ' at ' + ip4_addr + ':' + ip4_port);
+        sensor = new Sensor(mac_addr, ip4_addr, ip4_port, server);
+        this._sensors[mac_addr] = sensor;
+    } else {
+        sensor.updateAddress(ip4_addr, ip4_port);
+    }
+    return sensor;
+};
+
+// find the sensor object based on the partial id or returns undefined
+SalIO.prototype._getSensor = async function (sensorId) {
+    sensorId = '' + sensorId;
+    let sensor;
+
+    if(sensorId.length < 4 || sensorId.length > 12) return undefined;
+
+    // autocomplete the sensorId and find the sensor object
+    if (sensorId.length === 12) {
+        sensor = this._sensors[sensorId];
+    } else { // try to guess the rest of the id
+        for (var mac_addr in this._sensors) { // pick the first match
+            if (mac_addr.endsWith(sensorId)) {
+                sensorId = mac_addr;
+                sensor = this._sensors[sensorId];
+            }
+        }
+    }
+
+    // if couldn't find a live sensor
+    if (!sensor) return undefined;
```

Comment:
Should this throw an error instead?

Reply:
Oh, this service started out as a copy of roboscape, so that part was just copied over. Error here might be ok, but I'd have to look further to make sure.

Reply:
Ok, I've looked through it and decided that would be for the best. Old way wasn't updating `error` and would just return false. Now invalid or nonexistant device id sets `error`.